## BIO594 Final Project Proposal 

Jill Ashey

## Title: Comparison of RNA-seq alignment tools in non-model organisms

### Background 

Sequence alignment is an essential part in analyzing RNA-seq data. Many different alignment tools have been developed for this purpose; however, the sheer number of alignment tools makes it difficult to decide which tool might work best for a specific data set. This issue becomes even more complex when using these tools with non-model organisms, as the necessary genomic information to properly use these tools may not be available for these organisms. In order to evaluate what tools might be optimal for non-model organsism, it is essential to compare and assess how these tools perform in non-model organisms. 

For my project, I will compare three RNA-seq alignment tools using two coral species, *Acropora cervicornis* and *Pocillopora acuta*. The specific tools to be utilized are STAR (alignment against genome), Bowtie2 (alignment against transcriptome), and Kallisto (pseudoalignment); these were chosen in order to assess alignment quality with limited genomic information. 

### Methods 

The data I will be using was generated by F. Seneca during sediment stress experiments in Hawai'i and Florida in 2015-2016. He and his collaborators did the experiments and the RNA extraction and sequencing. 

#### Experiments

##### Hawai'i

*P. acuta* was exposed to varying levels of sediment (0, 40, and 400 mg/L) for up to 7 days. The sediment type was unsterilized terrigenous red dirt. On day 4 and 7, corals were collected from their treatments and preserved in liquid nitrogen. 

##### Florida 

*A. cervicornis* was exposed to varying levels of sediment (0, 10, 30, and 300 mg/L) for 30 days. The sediment type was sterilized coral rubble white sediment. After 30 days, corals were collected from their treatments and preserved in liquid nitrogen. 

#### RNA extraction and sequencing

RNA was extracted using the Direct-zol RNA MiniPrep kit and quantified using a Qubit. mRNA libraries were prepared using the NeoPrep kit. RNA was sequenced on a HiSeq 50 cycle single read sequencing v4 by the High Throughput Genomics Core Facility at the University of Utah. 

#### Bioinformatic analysis 

Raw sequence reads in fastq format are on the URI Andromeda server (add path). Genomic information for *P. acuta* was obtained from [here](http://cyanophora.rutgers.edu/Pocillopora_acuta/). Genomic information for *A. cervicornis* was obtained from [here](https://usegalaxy.org/u/skitch/h/acervicornis-genome) (Baums lab). 

For my analysis, I will complete the following workflow: 

![](https://raw.githubusercontent.com/jpuritz/BIO_594_2022/main/Exercises/course_project/JAshey/BIO594_FinalProject_Workflow.jpg)




